### DISCUSSION TITLE
[8 GPU and Specialized Hardware - part2]sch.annotate的作用和工作原理

### DISCUSSION THREAD

**Discussion by hbsun2113** [mlc-ai/mlc-zh#117] [Q&A] [answered]
在[第八节](https://github.com/mlc-ai/notebooks/blob/main/8_GPU_and_Specialized_Hardware_part2.ipynb)课程中，提到可以使用`sch.annotate(i, "pragma_import_llvm", tmm_kernel())`来替代`tvm.tir.TensorIntrin.register("tmm16", tmm16_desc, tmm16_impl)`。我的理解是对于block_mm内部的计算逻辑使用`tmm_kernel`而不是`tmm16_impl`。但是通过`sch.mod.show()`我发现block_mm的计算逻辑仍然使用了`T.call_extern("tmm16"...`。请问这里`sch.annotate`对实际的运行会产生什么影响？
![image](https://user-images.githubusercontent.com/27994442/184529046-37ce0425-740d-44d9-af89-a03769416069.png)

文档没有说明`sch.annotate`对于运行的影响和原理：
![image](https://user-images.githubusercontent.com/27994442/184529096-252d99a9-c21a-4316-b4b1-a3b418341f89.png)

**Comment by Hzfengsy** [comment, marked as answer]
可能你对这部分有一点误解。`sch.annotate` 并不能“替代” `tvm.tir.TensorIntrin.register("tmm16", tmm16_desc, tmm16_impl)`。这两者是需要同时作用才能工作的。
这里的`sch.annotate`是给定程序一个`tmm16`的实现（这个函数在是自己实现的，在编译器中找不到所以需要给定程序），配合Tensorize一起使用

**Reply by hbsun2113** [reply]
谢谢！按照您所说这里应该只给出了一个完整的流程。但notes里是说alternative，我理解是给出了两个都可以跑通的流程，现在看来本意是说拿micro-kernal作为一个例子说明？
![image](https://user-images.githubusercontent.com/27994442/184529638-27cafb15-9905-4165-90dd-ab4d1f9a7528.png)
那这里翻译为“或者”是否不太准确？
![image](https://user-images.githubusercontent.com/27994442/184529678-b5d75966-0377-46b7-97d1-14448e012dc7.png)

**Reply by Hzfengsy** [reply]
`alternative`，指的是“The downstream compilation step can easily map the implementation to an instruction that implements the operation.” 和 “we can map tmm16 to a micro-kernel that implements this tensorized computation.”的“或者”，即：
1. 如果下层编译器有这个"tmm" intrinsic，那我们就不用写mirco-kernel；
2. 如果下层编译器没有这个指令，那么我们需要写mirco-kernel。

**Reply by hbsun2113** [reply]
明白了。还有一个问题，我没有看到`tmm16`和`tmm_kernel`之间的映射，请问`sch.annotate`是如何工作的，使得tvm可以将`tmm16`的实现替换为`tmm_kernel`？
这里没有出现tmm16：sch.annotate(i, "pragma_import_llvm", tmm_kernel())。

**Reply by Hzfengsy** [reply]
`tmm_kernel`是一个Python函数，返回的是`tmm16`的llvm IR代码。

**Reply by hbsun2113** [reply]
请问哪里有指明`tmm_kernel`返回的是`tmm16`的代码呢？

**Reply by Hzfengsy** [reply]
`tmm_kernel`的实现 `extern "C" int tmm16(...)`

**Reply by hbsun2113** [reply]
明白了，感谢！
